### PR TITLE
feature: posting patient invoices

### PR DESCRIPTION
**Pull request by jniles**
This PR adds the ability to post patient invoices (`/sales`) to the posting journal.  It supports and tests posting sales with billing services and subsidies.  As discussed in #343, it will throw errors if there are more subsidies than one.

Posting patient invoices is a fairly complex operation.  It involves factoring in Billing Services, Subsidies, previous caution payments made by the debtor, and the actual sale items.

The Billing Services and Subsides are pre-processed in the `POST /sales` route, and written to the table `sale_billing_service` and `sale_subsidy` respectively.  There are some recommendations for how to improve the `/sales` API in the comments of the `finance/patientInvoice.js` controller.  The most important recommendations are:
1. No need to keep `debit`/`credit` fields in the `sale_item` table.  This can be renamed something a bit more meaningful like `total_price` or `final_price`.
2. No need to ship large objects for `billingServices` and `subsidies`, since we will look up all information based on the `id`.  I would recommend the API accept a sale object like this in the future:
   
   ``` js
   {
     description : "Some Sale"
     date : "2016-01-23",
   
     /* other properties ... */
   
     items : [ /* a list of sale item objects */ ],
     billingServices : [ /* a list of billing service ids */ ],
     subsidies : [ /* a list of subsidy ids. If #343 is adopted, this could be a single subsidy */ ]
   }
   ```
3. `description` should be `NOT NULL`.  See #347 for progress on that.
4. The totals calculation (`quantity` \* `transaction_price`) should be done on the server in MySQL.

Because of the pre-processing done by Billing Services and Subsidies, the total cost to the patient (factoring in Billing Services (additional fees) and Subsidies (additional reductions)) is already calculated and stored in the `sale.cost` column.  Posting only has to think about previous cautions to determine how much the debtor should pay.

The posting work is done in an SQL procedure - `PostPatientInvoice()`.  This procedure does the following operations:
1. Loops through any previous caution payments, assigning part of the cost of the sale (up to the caution's total amount) to the caution using the `reference_uuid`.  If the caution is more than the sale's total cost, we assign the entire amount to the caution and exit the loop.
2. After the loop, if there is any left over amount (debtor's bill was greater than the sum of all his/her previous cautions), we write it in a single line in the `posting_journal`, with a NULL `reference_uuid`.
3. Lines for the `sale_item`s are copied into the posting journal next.
4. Lines for `sale_subsidy`s are copied.
5. Lines for `sale_billing_service`s are copied.

This completes the posting process.

Unfortunately, I had to include a hack to unset the session variables created during journal operations  since we do not handle session variables properly in MySQL (see `journal/core.js`'s `cleanup()` function).  There will be an issue created about this after this PR and I'll update here.

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process
and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by jniles** [comment]
~~@sfount finally got my selenium up and running after a node update clobbered it, and I think the e2e tests have some issues.  I'll investigate this evening.~~

Edit: round two seems to have worked fine.  Let me know what you find on your machine.

**Comment by sfount** [comment]
@jniles okay, I will pull this and run tests to double check it's not your
instance.

File: server/models/test/sample.sql [hunk 1]
```diff
@@ -0,0 +1,29 @@
+
```
Comment (sfount) [inline server/models/test/sample.sql]: @jniles Did you intend to commit this sample.sql?

Comment (jniles) [inline server/models/test/sample.sql]: Nope!  Hang on, I'll remove it.

**Comment by sfount** [comment]
Pull request LGTM - excellent starting point for having very high standards (safe) yet complex server side posting that can be built on.

**Comment by jniles** [comment]
@sfount removed the errant `sample.sql` file.

**Comment by jniles** [comment]
@sfount could I get another review of this code?

File: server/test/api/helpers.js [hunk 2]
```diff
@@ -31,6 +31,20 @@ exports.login = function login(agent) {
...
 
+/**
+ * Clones the object and removes the field, to test if the field is required
+ * and the server error codes
+ *
+ * @param {Object} object - any valid JS object
+ * @param {String} field - a key on the passed in object
+ * @returns {Object} clone - the copied object missing the propertay
+ */
+exports.mask = function mask(object, field) {
```
Comment (sfount) [inline server/test/api/helpers.js:42]: Good idea. This should help reduce boilerplate.

**Comment by sfount** [comment]
Code still LGTM. My machine passes all integration tests but has one failing end to end test on the cash boxes module. I'll run it again to verify. 

Edit: 
screenshot 2016-05-02 at 09 37 09

**Comment by jniles** [comment]
@sfount, thanks for the update.  I'll try to track this down.

**Comment by jniles** [comment]
Hmm...  cannot reproduce on my machine.

e2e

**Comment by sfount** [comment]
@jniles did you re-base with master first? If you did then we can assume mine is a fluke.

**Comment by jniles** [comment]
Yessir!

**Pull request merged**